### ISSUE TITLE
Roundup: [oss-security] util-linux 2.29.2 fixes CVE-2017-2616

### ISSUE THREAD

**Issue by grahamc** [NixOS/security#29]
Here is a vulnerability from the oss-security mailing list
for [Vulnerability Roundup 26](https://github.com/NixOS/nixpkgs/issues/24161).

# Instructions:
## Identification
Identify if we have the software, in 16.09, 17.03, and unstable.
Then determine if we are vulnerable, and make a comment with
your findings. It can also be helpful to specify if you think there is
a patch, or if it can be fixed via a general update.

Example:

```
unstable: we are not vulnerable (link to the package)
17.03: we are vulnerable (link to the package)
16.09: we don't have it packaged
```

**IMPORTANT**: If you believe there are possibly related issues, bring
them up on the _parent_ issue!

## Patching
Start by commenting on this issue saying you're working on a patch.
This way, we don't duplicate work.

If you open a pull request, tag this issue _and_ the master issue
for the roundup.

If you commit the patch directly to a branch, please leave a comment
on this issue with the branch and the commit hash, example:

```
fixed:

release-16.09: abc123
```

## Upon Completion ...
 - [ ] Update Graham's database

## Info

_Triage Indicator:_

```
-needs-triage +roundup26 thread:0000000000003dc6 # [oss-security] util-linux 2.29.2 fixes CVE-2017-2616
```

 - File Search: https://search.nix.gsc.io/?q=util-linux&i=fosho&repos=nixos-nixpkgs
 - GitHub Search: https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=util-linux+in%3Apath&type=Code

---

##### Thu, 23 Feb 2017 08:46:30 +0100 Marcus Meissner <meissner-at-suse.de>, `20170223074630.GB26098@suse.de`

    Hi,
    
    util-linux 2.29.2 fixes CVE-2017-2616, a race condition which allowed local users
    to kill other processes.
    
    https://www.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29.2-ReleaseNotes
    
    "
      It is possible for any local user to send SIGKILL to other processes with root
      privileges.  To exploit this, the user must be able to perform su with a
      successful login.  SIGKILL can only be sent to processes which were executed
      after the su process.  It is not possible to send SIGKILL to processes which
      were already running.
    "
    
    Root cause of the flaw that a regular exit of the child process and the su ctrl-c kill of the
    child PID could race and so you would be able to later started process with this specific PID.
    
    The fix is here:
    https://github.com/karelzak/util-linux/commit/dffab154d29a288aa171ff50263ecc8f2e14a891
    
    Ciao, Marcus

---

##### Thu, 23 Feb 2017 10:40:54 +0100 Hanno Böck <hanno-at-hboeck.de>, `20170223104054.42060545@pc1`

    Hi,
    
    On Thu, 23 Feb 2017 08:46:30 +0100
    Marcus Meissner <meissner@suse.de> wrote:
    
    > util-linux 2.29.2 fixes CVE-2017-2616, a race condition which allowed
    > local users to kill other processes.
    
    I just reported this in Gentoo [1], yet I was informed that we're not
    using su from util-linux, but from shadow. So depending on the
    distribution you may not use this implementation of su.
    
    I haven't digged deeper into this, can you say if this issue is
    generic enough to be expected in other implementations as well? (Not
    sure if the implementations of su in shadow and util-linux share a
    common codebase, seems to be quite old stuff.)
    
    [1] https://bugs.gentoo.org/show_bug.cgi?id=610664
    -- 
    Hanno Böck
    https://hboeck.de/
    
    mail/jabber: hanno@hboeck.de
    GPG: FE73757FA60E4E21B937579FA5880072BBB51E42

---

##### Thu, 23 Feb 2017 11:01:48 +0100 Marcus Meissner <meissner-at-suse.de>, `20170223100148.GA23394@suse.de`

    On Thu, Feb 23, 2017 at 10:40:54AM +0100, Hanno Böck wrote:
    > Hi,
    > 
    > On Thu, 23 Feb 2017 08:46:30 +0100
    > Marcus Meissner <meissner@suse.de> wrote:
    > 
    > > util-linux 2.29.2 fixes CVE-2017-2616, a race condition which allowed
    > > local users to kill other processes.
    > 
    > I just reported this in Gentoo [1], yet I was informed that we're not
    > using su from util-linux, but from shadow. So depending on the
    > distribution you may not use this implementation of su.
    > 
    > I haven't digged deeper into this, can you say if this issue is
    > generic enough to be expected in other implementations as well? (Not
    > sure if the implementations of su in shadow and util-linux share a
    > common codebase, seems to be quite old stuff.)
    > 
    > [1] https://bugs.gentoo.org/show_bug.cgi?id=610664
    
    coreutils uses the same su.c codebase, so it is also affected.
    
    
    Looking at shadow su.c code, it calls waitpid ... and does not handle
    the pid_child exiting case after calling waitpid.
    
    So I would think it is affected without digging deeper.
    
    Ciao, Marcus

---

##### Thu, 23 Feb 2017 07:56:51 -0500 Assaf Gordon <assafgordon-at-gmail.com>, `D7FFF88A-2AA9-432A-9830-0DF662E32524@gmail.com`

    Hello,
    
    > On Feb 23, 2017, at 05:01, Marcus Meissner <meissner@suse.de> wrote:
    > 
    > On Thu, Feb 23, 2017 at 10:40:54AM +0100, Hanno Böck wrote:
    >> 
    >>> util-linux 2.29.2 fixes CVE-2017-2616, a race condition which allowed
    >>> local users to kill other processes.
    > 
    > coreutils uses the same su.c codebase, so it is also affected.
    > 
    
    GNU Coreutils stopped installing 'su' by default in 2007,
    and completely removed 'su' (including the 'su.c' source file)
    in 2012.
    
    See:
    https://git.savannah.gnu.org/cgit/coreutils.git/commit/?id=928dd737
    
    regards,
     - assaf

---

##### Thu, 23 Feb 2017 17:08:48 +0100 Hanno Böck <hanno-at-hboeck.de>, `20170223170848.17d43fb3@pc1`

    On Thu, 23 Feb 2017 07:56:51 -0500
    Assaf Gordon <assafgordon@gmail.com> wrote:
    
    > GNU Coreutils stopped installing 'su' by default in 2007,
    > and completely removed 'su' (including the 'su.c' source file)
    > in 2012.
    
    That's good to know, so now there are only 2 competing versions of su
    instead of 3 in major packages :-)
    
    Anyone have a good idea who is using shadow vs. util-linux su? Do they
    have specific advantages/disadvantages, would it be reasonable to try
    to get all distros to use them same one?
    
    -- 
    Hanno Böck
    https://hboeck.de/
    
    mail/jabber: hanno@hboeck.de
    GPG: FE73757FA60E4E21B937579FA5880072BBB51E42

---

##### Thu, 23 Feb 2017 17:17:28 +0100 Emilio Pozuelo Monfort <pochu-at-debian.org>, `89dcadc4-9895-a6e2-36ed-ac3195137978@debian.org`

    On 23/02/17 17:08, Hanno Böck wrote:
    > On Thu, 23 Feb 2017 07:56:51 -0500
    > Assaf Gordon <assafgordon@gmail.com> wrote:
    > 
    >> GNU Coreutils stopped installing 'su' by default in 2007,
    >> and completely removed 'su' (including the 'su.c' source file)
    >> in 2012.
    > 
    > That's good to know, so now there are only 2 competing versions of su
    > instead of 3 in major packages :-)
    > 
    > Anyone have a good idea who is using shadow vs. util-linux su?
    
    Debian is using shadow's, fwiw.
    
    Emilio

---

##### Thu, 23 Feb 2017 10:20:28 -0600 "Serge E. Hallyn" <serge-at-hallyn.com>, `20170223162028.GA14624@mail.hallyn.com`

    On Thu, Feb 23, 2017 at 05:08:48PM +0100, Hanno Böck wrote:
    > On Thu, 23 Feb 2017 07:56:51 -0500
    > Assaf Gordon <assafgordon@gmail.com> wrote:
    > 
    > > GNU Coreutils stopped installing 'su' by default in 2007,
    > > and completely removed 'su' (including the 'su.c' source file)
    > > in 2012.
    > 
    > That's good to know, so now there are only 2 competing versions of su
    > instead of 3 in major packages :-)
    > 
    > Anyone have a good idea who is using shadow vs. util-linux su? Do they
    > have specific advantages/disadvantages, would it be reasonable to try
    > to get all distros to use them same one?
    
    There is a debian bug to switch over to the util-linux version of su.

---

##### Thu, 23 Feb 2017 17:31:42 +0100 Bálint Réczey <balint-at-balintreczey.hu>, `CAK0OdpxiQO7Mt6TCSn-92D+nwnj3iJCwU4OTiuZ1=Kn5Rx_-xg@mail.gmail.com`

    Hi,
    
    2017-02-23 17:08 GMT+01:00 Hanno Böck <hanno@hboeck.de>:
    > On Thu, 23 Feb 2017 07:56:51 -0500
    > Assaf Gordon <assafgordon@gmail.com> wrote:
    >
    >> GNU Coreutils stopped installing 'su' by default in 2007,
    >> and completely removed 'su' (including the 'su.c' source file)
    >> in 2012.
    >
    > That's good to know, so now there are only 2 competing versions of su
    > instead of 3 in major packages :-)
    >
    > Anyone have a good idea who is using shadow vs. util-linux su? Do they
    > have specific advantages/disadvantages, would it be reasonable to try
    > to get all distros to use them same one?
    
    In Debian we are looking into switching to util-linux from shadow for
    commands provided by both packages:
    https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=833256
    
    Cheers,
    Balint

---

##### Thu, 23 Feb 2017 11:51:39 -0500 Leo Famulari <leo-at-famulari.name>, `20170223165139.GA11848@jasmine`

    On Thu, Feb 23, 2017 at 05:08:48PM +0100, Hanno Böck wrote:
    > Anyone have a good idea who is using shadow vs. util-linux su? Do they
    > have specific advantages/disadvantages, would it be reasonable to try
    > to get all distros to use them same one?
    
    Debian uses shadow, but is discussing a switch to util-linux:
    https://packages.debian.org/sid/login
    https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=833256
    
    GuixSD is using shadow:
    https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/system.scm#n579
    
    I'm also interested to learn the pros and cons of the two
    implementations.

<details><summary>signature.asc</summary>

    -----BEGIN PGP SIGNATURE-----
    
    iQIzBAABCAAdFiEEsFFZSPHn08G5gDigJkb6MLrKfwgFAlivExgACgkQJkb6MLrK
    fwidNg/+PbdtVzcQgcYAIw3LLE6La+HPmIfWcRMja0QCEb47ruScq2AwKn2HNQQw
    YTesr4NWzNhUs5esctwjXqlo5V9C5OtKJUN9So/S12qCjc7DN0g7e8fkSgTrqbU0
    v1zT72/XVl1gIlS62ro0z9MNgsC9cy8zJUp/KVcllIE69u0Zk4ZuprJjbnZfhFG6
    OYedkmUgnxZ8PfaUuxCGnwewI3WMpd19NcMWrfmXlcGkiAkPTB6d9t0GyHMBdjjM
    QGoDj+lfzFdKmSpbGDdyMIpu/VHsdlVs18KAO9JzXMWe5YNJjv47ex3uAytB5VQ0
    VScw957+/ptn6Qeru67JpuZLTZTh5/6Q2YToJSGyzbXqAWxzlz7OvYZhpgW14fON
    EqDc06iGlogFwKBeR8UFEtPLWWNYF+JThMATbVrmJ5rq35rmaTZc4AVQ2PAyYaOh
    e+SUMtvG7Mn1lkTntQrRenZ1oaQ/UTYvnveF5JkEWFg2duGapvp6pwu7dXgHE729
    vk1qRN2lLjcnJe2kWghk/X0/IUpb0xMVm1C13OjCQipbIZeJjS5Nz/xcX5gDiErN
    yp0AdmImHwkDTnqzk+2VIyafPgs8iPK9EOCE2GW8yqwJbprB3Xs1lsUIBvRgjkOF
    9uWR94RvIJGXYJhwSPdhOfOFfoQqT3nkmesuXdKXdxMR5a5/NPA=
    =ahFy
    -----END PGP SIGNATURE-----
    

</details>

---

##### Thu, 23 Feb 2017 19:10:51 +0100 Tobias Stöckmann <tobias-at-stoeckmann.org>, `a3beb7ec-c3e0-6506-cbbc-18a7f92eeca4@stoeckmann.org`

    Hello,
    
    as I have discovered the issue, I can confirm that su of util-linux as
    well as shadow are affected.
    
    I have supplied patches to both maintainers and the shadow patch has
    been pushed now, too.
    
    The code shares a common origin but differs slightly, which in fact made
    the patch for shadow a bit trickier.
    
    But as the code bases are so close to each other, I don't think that it
    takes another CVE-ID for this one.
    

<details><summary>signature.asc</summary>

    -----BEGIN PGP SIGNATURE-----
    Version: GnuPG v2
    
    iQIcBAEBCAAGBQJYryW0AAoJEH20cPNbC4sYSYUQALMTMIFkBwpaRP//n3kEkcic
    HngrYo6wAsEwkjHorSNNzplqSAqySSCN1M3XwhfyjnFktDjrVwCOQZ5SzeK5lF2+
    g3ZLarWxyKe9Q0waOSU0QW/N1KPDgM6BlIUbRZueuHX0BOQ6m1fn9HfPukqPkeOI
    Xsw7DInUSTf5udtdkGYYMYyzq7w/8dzK66lMLKe2A4cS8WvQZK1RkDPXuGH2a4D0
    Q0j8H+EspbxzbCcSPH+XQIVuTCUqa1z6ky8cvWyFu5VsW7nqVz9HWLeAWRAgrNPo
    HB5WYA8f2H3pUVp9S5OCjBEu+0t3kmTca+xZpXeZ4lEW/3Ho3wIpOZYSTXuhOj3x
    c/LMKCjIfaLBV9xjZt4fpB3NmenfNPs9Fs2jlmI78QqqEv8FQPMNs/VbnHnzlRqW
    W42Iyvquov6lUh7qajal0qoH1isNO/JFyCq94ZrhARX2M2kcS4CUhFUl5Pt9W39D
    fBsi1/PPnNDuVGfflBXarzV5PJrJFk0Njw76u0LWSqJaJZxuZGn/uU+CFVgjFJhL
    E16M37Wwcpe0Odh+m4v4oOREpmOEfl/bDLaIH23Nb+ZmK3UvPf8zJQjlLnQe9A/C
    UXXqrKca1vtvDaxLeI2KdOguc5/pJ+xRvogluZkOg/AY51nnVjyHyH+iu10gyXAp
    +bF6GsSLV23YGxCt+06P
    =rIX4
    -----END PGP SIGNATURE-----
    

</details>

---

**Comment by 7c6f434c** [comment]
Status:
`master`, `release-17.03`: fresh version already
`release-16.09`: vulnerable version, but `su` from `shadow` is used by default…

And `shadow` is also vulnerable, reference in the parent issue.

**Comment by 7c6f434c** [comment]
An update is costly, better to push together with the `shadow` update

**Comment by fpletz** [comment]
This has been fixed: 9d14ea4295d935dce73c16f3ceb025cbb2443c5a
